File: backend/app/schemas/ai.py
from datetime import date
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class AIChatMessage(BaseModel):
    role: str = Field(..., description="'user' or 'assistant' or 'system'")
    content: str


class AIChatRequest(BaseModel):
    message: str = Field(..., description="Câu hỏi hoặc yêu cầu của người dùng")
    history: List[AIChatMessage] = Field(
        default=[], description="Lịch sử hội thoại trước đó"
    )
    include_financial_context: bool = Field(
        default=True,
        description="Đính kèm bối cảnh số dư live, nợ đến hạn và chi tiêu vào prompt",
    )


class AIChatResponse(BaseModel):
    reply: str
    suggested_followups: List[str] = []
    insights: Optional[Dict[str, Any]] = None


class AIParsedTransactionItem(BaseModel):
    transaction_date: date
    post_date: Optional[date] = None
    raw_description: str
    amount: Decimal
    fee: Decimal = Decimal("0.00")
    total_amount: Decimal
    original_amount: Optional[Decimal] = None
    original_currency: Optional[str] = "VND"
    exchange_rate: Optional[Decimal] = Decimal("1.00")
    category_hint: Optional[str] = None
    transaction_type_hint: Optional[str] = "PURCHASE"


class AIParsedStatementResult(BaseModel):
    bank_detected: str
    account_number_hint: Optional[str] = None
    statement_date: date
    payment_due_date: date
    credit_limit: Decimal
    previous_balance: Decimal = Decimal("0.00")
    purchases_amount: Decimal = Decimal("0.00")
    payments_received: Decimal = Decimal("0.00")
    statement_balance: Decimal
    minimum_payment: Decimal
    earned_reward_points: Optional[Decimal] = None
    transactions: List[AIParsedTransactionItem] = []
    parsing_notes: Optional[str] = None


class AIPdfExtractionResponse(BaseModel):
    success: bool
    filename: str
    data: Optional[AIParsedStatementResult] = None
    error: Optional[str] = None

File: frontend/src/types/ai.ts
import { TransactionType } from "./transaction";

export interface AITransactionDraft {
  action_type?: string;
  account_id?: string;
  account_name?: string;
  account_bank_name?: string;
  transaction_type: TransactionType;
  amount: number;
  fee?: number;
  total_amount?: number;
  transaction_date: string;
  post_date?: string;
  category_id?: string;
  category_name?: string;
  parent_category_name?: string;
  merchant_name?: string;
  raw_description?: string;
  note?: string;
  transfer_to_account_id?: string;
  transfer_to_account_name?: string;
  confidence_score?: number;
}

export interface AIChatMessage {
  role: "user" | "assistant" | "system";
  content: string;
  action?: string;
  transaction_draft?: AITransactionDraft;
  draft_status?: "pending" | "confirmed" | "cancelled";
}

export interface AIChatRequest {
  message: string;
  history?: AIChatMessage[];
  include_financial_context?: boolean;
}

export interface AIChatResponse {
  reply: string;
  suggested_followups?: string[];
  insights?: Record<string, any>;
  action?: string;
  transaction_draft?: AITransactionDraft;
}


export interface AIParsedTransaction {
  transaction_date: string;
  post_date?: string;
  raw_description: string;
  amount: number;
  fee?: number;
  total_amount: number;
  original_amount?: number;
  original_currency?: string;
  exchange_rate?: number;
  category_hint?: string;
  transaction_type_hint?: string;
}

export interface AIParsedStatement {
  bank_detected: string;
  account_number_hint?: string;
  statement_date: string;
  payment_due_date: string;
  credit_limit: number;
  previous_balance: number;
  purchases_amount: number;
  payments_received: number;
  statement_balance: number;
  minimum_payment: number;
  earned_reward_points?: number;
  transactions: AIParsedTransaction[];
  parsing_notes?: string;
}

export interface AIPdfExtractionResponse {
  success: boolean;
  filename: string;
  data?: AIParsedStatement;
  error?: string;
}
